src/librbd/Utils.h

```diff
@@ -152,6 +203,33 @@ Context *create_context_callback(T *obj) {
   return new detail::C_StateCallbackAdapter<T, MF, destroy>(obj);
 }
 
+//for reference counting objects
+template <typename T, void(T::*MF)(int) = &T::complete>
+Context *create_context_callback(T *obj, RefCountedPtr refptr) {
+  return new detail::C_RefCallbackAdapter<T, MF>(obj, refptr);
+}
+
+template <typename T, Context*(T::*MF)(int*), bool destroy=true>
+Context *create_context_callback(T *obj, RefCountedPtr refptr) {
+  return new detail::C_RefStateCallbackAdapter<T, MF, destroy>(obj, refptr);
+}
+
+//for objects that don't inherit from RefCountedObj, to handle unit tests
+template <typename T, void(T::*MF)(int) = &T::complete, typename R>
+Context *create_context_callback(T *obj, R *refptr) {
```

Comment:
For safety purposes, you might want to disable these two methods if `R` derives from `RefCounterPtr`.

`typename std::enable_if<not std::is_base_of<RefCountedPtr, R>::value, Context*>::type
crete_context_callback(T *obj, R *refptr) {
`

Reply:
Done